include/roboteam_ai/control/BBTrajectories/BBTrajectory1D.h

```diff
@@ -0,0 +1,124 @@
+        * @param initialPos Current position
+        * @param initialVel Current velocity
+        * @param finalPos Final position
+        * @param maxVel Maximum allowed velocity (absolute)
+        * @param maxAcc maximum allowed acceleration/deceleration
+        */
+        void generateTrajectory(num initialPos, num initialVel, num finalPos, num maxVel, num maxAcc) noexcept ;
+        BBTrajectory1D(num initialPos, num initialVel,num finalPos, num maxVel, num maxAcc) noexcept;
+        BBTrajectory1D() noexcept = default;
+    private:
+
+        /**
+         * Computes the position where we would end if we would initiate a maximal break given current state
+         * @param pos Current position
+         * @param vel Current velocity
+         * @param maxAcc Maximum allowed acceleration/deceleration
+         * @return final position (with zero velocity)
+         */
+        num fullBrakePos(num pos, num vel, num accMax) noexcept ;
+        /**
+         * Computes the position where we would end if we first accelerate to a target velocity and then immediately decelerate to 0
+         * @param pos0 Current position
+         * @param vel0 Current velocity
+         * @param vel1 Target velocity
+         * @param accMax Maximum allowed acceleration/deceleration
+         * @return final position (with zero velocity)
+         */
+        num accelerateBrakePos(num pos0, num vel0, num vel1, num accMax) noexcept;
+        /**
+         * Generates a time-optimal triangular profile (accelerating and then breaking)
+         * @param initialPos Current position
+         * @param initialVel Current velocity
+         * @param finalPos Desired final psition
+         * @param maxAcc Maximum allowed acceleration/deceleration
+         * @param invertedSign Specifies whether or not to invert the sign of velocity / acceleration. This is more of a techniality
+         */
+        void triangularProfile(num initialPos, num initialVel, num finalPos, num maxAcc, bool invertedSign) noexcept;
+        /**
+         * Generates a time-optimal trapezoidal profile (accelerating, coasting, then breaking)
+         * @param initialPos Current position
+         * @param initialVel Current velocity
+         * @param maxVel Coasting velocity
+         * @param finalPos Desired final position
+         * @param maxAcc Maximum allowed acceleration/deceleration
+         */
+        void trapezoidalProfile(num initialPos, num initialVel, num finalPos, num maxVel, num maxAcc) noexcept;
+        /**
+         * Updates a part of the trajectory
+         * @param index which part to update
+         * @param tEnd time spent on trajectory segment
+         * @param acc acceleration on segment
+         * @param vel initial velocity of segment
+         * @param pos initial position of segment
+         */
+        void updatePart(int index, num tEnd, num acc, num vel, num pos) noexcept;
+
+        part parts[3];
```

Comment:
std::array<part, 3>
```suggestion
        std::array<part, 3> parts;
```

Reply:
Actually got a question about this since this class might not necessary use all 3 parts. I now use the numParts variable to keep track how many parts there are (which can be 1,2 or 3) but this is impossible to know beforehand. I *could* use std::vector but I would prefer to have heap allocations since this calculation really needs to be fast at some point. For now I decided to leave optimizations for later, but if you have good suggestions on how to handle this I would be open to them.